lib/ansible/utils/collection_loader.py

```diff
@@ -0,0 +1,305 @@
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+import os.path
+import pkgutil
+import re
+import sys
+
+from importlib import import_module
+from types import ModuleType
+
+from ansible import constants as C
+from ansible.module_utils._text import to_native
+from ansible.module_utils.six import iteritems, string_types
+
+_SYNTHETIC_PACKAGES = {
+    'ansible_collections.ansible': dict(type='pkg_only'),
+    'ansible_collections.ansible.core': dict(type='pkg_only'),
+    'ansible_collections.ansible.core.plugins': dict(type='map', map='ansible.plugins'),
+    'ansible_collections.ansible.core.plugins.module_utils': dict(type='map', map='ansible.module_utils', graft=True),
+    'ansible_collections.ansible.core.plugins.modules': dict(type='flatmap', flatmap='ansible.modules', graft=True),
+}
+
+# TODO: tighten this up to subset Python identifier requirements
+_collection_role_name_re = re.compile(r'^(\w+)\.(\w+)\.(\w+)$')
+
+# FIXME: exception handling/error logging
+class AnsibleCollectionLoader(object):
+    def __init__(self):
+        self._configured_paths = C.config.get_config_value('INSTALLED_CONTENT_ROOTS')
+
+        if isinstance(self._configured_paths, string_types):
+            self._configured_paths = [self._configured_paths]
+        elif self._configured_paths is None:
+            self._configured_paths = []
+
+        # expand any placeholders in configured paths
+        self._configured_paths = [os.path.expanduser(p) for p in self._configured_paths]
+
+        self._playbook_paths = []
+        # pre-inject grafted package maps so we can force them to use the right loader instead of potentially delegating to a "normal" loader
+        for p in (p for p in iteritems(_SYNTHETIC_PACKAGES) if p[1].get('graft')):
+            pkg_name = p[0]
+            pkg_def = p[1]
+
+            newmod = ModuleType(pkg_name)
+            newmod.__package__ = pkg_name
+            newmod.__file__ = '<ansible_synthetic_collection_package>'
+            pkg_type = pkg_def.get('type')
+
+            # TODO: need to rethink map style so we can just delegate all the loading
+
+            if pkg_type == 'flatmap':
+                newmod.__loader__ = AnsibleFlatMapLoader(import_module(pkg_def['flatmap']))
+            newmod.__path__ = []
+
+            sys.modules[pkg_name] = newmod
+
+
+    @property
+    def _collection_paths(self):
+        return self._playbook_paths + self._configured_paths
+
+    def set_playbook_paths(self, playbook_paths):
+        if isinstance(playbook_paths, string_types):
+            playbook_paths = [playbook_paths]
+
+        added_paths = set()
+
+        # de-dupe and ensure the paths are native strings (Python seems to do this for package paths etc, so assume it's safe)
+        self._playbook_paths = [to_native(p) for p in playbook_paths if not (p in added_paths or added_paths.add(p))]
```

Comment:
I'm pretty sure that there's no need to check the value in set
```suggestion
        self._playbook_paths = [to_native(p) for p in playbook_paths if not added_paths.add(p)]
```

Reply:
Is that what you're trying to do?